frontend/javascripts/dashboard/folders/metadata_table.tsx

```diff
@@ -0,0 +1,414 @@
+  );
+  const [focusedRow, setFocusedRow] = useState<number | null>(null);
+  const [isSaving, setIsSaving] = useState<boolean>(false);
+  const [hasUnsavedChanges, setHasUnsavedChanges] = useState<boolean>(false);
+
+  const isFirstRenderingCycle = previousDatasetOrFolder == null;
+
+  // Always update local state when folder / dataset changes. Prior to that send pending updates.
+  // biome-ignore lint/correctness/useExhaustiveDependencies: Only execute this hook when underlying datasetOrFolder changes.
+  useEffect(() => {
+    if (isFirstRenderingCycle) {
+      // Skip first rendering cycle.
+      return;
+    }
+    const isSameName = datasetOrFolder.name === previousDatasetOrFolder.name;
+    const isSameType = isDataset(datasetOrFolder) === isDataset(previousDatasetOrFolder);
+    const isSameDatasetOrFolder = isSameName && isSameType;
+    const effectCancelSignal = { dontUpdateState: false };
+    if (!isSameDatasetOrFolder && hasUnsavedChanges) {
+      // Flush pending updates as the dataset or folder changed.
+      saveCurrentMetadata(
+        previousDatasetOrFolder,
+        metadata,
+        context,
+        effectCancelSignal,
+        setIsSaving,
+        setHasUnsavedChanges,
+      );
+    }
+    setMetadata(
+      datasetOrFolder.metadata?.map((entry, index) => ({ ...entry, index, error: null })) || [],
+    );
+    return () => {
+      effectCancelSignal.dontUpdateState = true;
+    };
+  }, [datasetOrFolder.name, isDataset(datasetOrFolder), datasetOrFolder.metadata]);
+
+  // Sent automatic debounced updates to the server when metadata changes.
+  // biome-ignore lint/correctness/useExhaustiveDependencies: Only update upon pending changes.
+  useEffect(() => {
+    const effectCancelSignal = { dontUpdateState: false };
+    if (hasUnsavedChanges) {
+      saveCurrentMetadataDebounced(
+        datasetOrFolder,
+        metadata,
+        context,
+        effectCancelSignal,
+        setIsSaving,
+        setHasUnsavedChanges,
+      );
+    }
+    return () => {
+      effectCancelSignal.dontUpdateState = true;
+    };
+  }, [metadata]);
+
+  const updateMetadataKey = (index: number, newPropName: string) => {
+    setMetadata((prev: IndexedMetadataEntries) => {
+      let error = null;
+      const entry = prev.find((prop) => prop.index === index);
+      if (!entry) {
+        return prev;
+      }
+      const maybeAlreadyExistingEntry = prev.find((prop) => prop.key === newPropName);
+      if (maybeAlreadyExistingEntry) {
+        error = `Property ${newPropName} already exists.`;
+      }
+      if (newPropName === "") {
+        error = "Property name cannot be empty.";
+      }
+      const detailsWithoutEditedEntry = prev.filter((prop) => prop.index !== index);
+      setHasUnsavedChanges(true);
+      return [
+        ...detailsWithoutEditedEntry,
+        {
+          ...entry,
+          key: newPropName,
+          error,
+        },
+      ];
```

Comment:
is there a reason for moving the item to the back? intuitively, I would have mapped the items and only update the one with the requested index (even though I'm still not sure I get the relevance of the index; isn't it redundant?).